### PR TITLE
Call zk.Sync() always before Get()

### PR DESCRIPTION
**Pull request by unakatsuo**
Zookeeper does not provide consistency for reading by its design. This rule seems to be applied to the read within a same session (connection).

For #143 & #149, the symptom and conditions are:
 
- Zookeeper is running in Standalone mode.
- There is one connection to the server and the write operation happens several times.
- The read occurs in the same connection as the write but the past result is returned.

[SyncRequestProcessor.java](https://github.com/apache/zookeeper/blob/release-3.4.6/src/java/main/org/apache/zookeeper/server/SyncRequestProcessor.java)
[FinalRequestProcessor.java](https://github.com/apache/zookeeper/blob/release-3.4.6/src/java/main/org/apache/zookeeper/server/FinalRequestProcessor.java)

- [x] Modify ``model/backend/zk.go``
- [x] Configure Zookeeper on multibox to multi-node mode.

### CODE REVIEW

**Comment by unakatsuo** [comment]
https://ci.openvdc.org/job/citest/job/zk-sync/5/consoleText

```
May 16 13:13:10 zookeeper zookeeper[681]: 2017-05-16 04:13:10,034 [myid:1] - WARN  [WorkerSender[myid=1]:QuorumCnxManager@382] - Cannot open channel to 3 at election address /10.0.100.12:3888
May 16 13:13:10 zookeeper zookeeper[681]: java.net.NoRouteToHostException: No route to host (Host unreachable)
May 16 13:13:10 zookeeper zookeeper[681]: at java.net.PlainSocketImpl.socketConnect(Native Method)
May 16 13:13:10 zookeeper zookeeper[681]: at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
May 16 13:13:10 zookeeper zookeeper[681]: at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
May 16 13:13:10 zookeeper zookeeper[681]: at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
May 16 13:13:10 zookeeper zookeeper[681]: at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
May 16 13:13:10 zookeeper zookeeper[681]: at java.net.Socket.connect(Socket.java:589)
May 16 13:13:10 zookeeper zookeeper[681]: at org.apache.zookeeper.server.quorum.QuorumCnxManager.connectOne(QuorumCnxManager.java:368)
May 16 13:13:10 zookeeper zookeeper[681]: at org.apache.zookeeper.server.quorum.QuorumCnxManager.toSend(QuorumCnxManager.java:341)
May 16 13:13:10 zookeeper zookeeper[681]: at org.apache.zookeeper.server.quorum.FastLeaderElection$Messenger$WorkerSender.process(FastLeaderElection.java:449)
May 16 13:13:10 zookeeper zookeeper[681]: at org.apache.zookeeper.server.quorum.FastLeaderElection$Messenger$WorkerSender.run(FastLeaderElection.java:430)
May 16 13:13:10 zookeeper zookeeper[681]: at java.lang.Thread.run(Thread.java:748)
```

**Comment by unakatsuo** [comment]
``srvr`` Result

```
=======> 10.0.100.10:2181
+ echo '=======> 10.0.100.10:2181'
+ echo srvr
+ nc 10.0.100.10 2181
Zookeeper version: 3.4.6-1569965, built on 02/20/2014 09:09 GMT
Latency min/avg/max: 0/1/9
Received: 72
Sent: 71
Connections: 3
Outstanding: 0
Zxid: 0x10000003e
Mode: follower
Node count: 12
=======> 10.0.100.11:2181
+ for addr in '${ZK_NODES[@]}'
+ echo '=======> 10.0.100.11:2181'
+ echo srvr
+ nc 10.0.100.11 2181
Zookeeper version: 3.4.6-1569965, built on 02/20/2014 09:09 GMT
Latency min/avg/max: 0/2/66
Received: 200
Sent: 202
Connections: 8
Outstanding: 0
Zxid: 0x10000003e
Mode: leader
Node count: 12
+ for addr in '${ZK_NODES[@]}'
+ echo '=======> 10.0.100.12:2181'
=======> 10.0.100.12:2181
+ echo srvr
+ nc 10.0.100.12 2181
Zookeeper version: 3.4.6-1569965, built on 02/20/2014 09:09 GMT
Latency min/avg/max: 0/4/18
Received: 22
Sent: 21
Connections: 1
Outstanding: 0
Zxid: 0x10000003e
Mode: follower
Node count: 12
```

File: ci/citest/acceptance-test/multibox/10.0.100.10-zookeeper/guestroot/var/lib/zookeeper/myid [hunk 1]
```diff
@@ -0,0 +1 @@
+1
```
Comment (Metallion) [inline ci/citest/acceptance-test/multibox/10.0.100.10-zookeeper/guestroot/var/lib/zookeeper/myid:1]: What is the purpose of this file?

I assume this is internally used by zookeeper as a node id?

**Review by Metallion — CHANGES_REQUESTED** [review]
If I understand correctly we are now running zookeeper on the `mesos-master` and `openvdc-scheduler` machines as well? The drawing in the illustrations folder will need to be updated to reflect this. Make sure to save it in multiline format as written in the README file located in the same illustrations folder.

Also, this means OpenVDC is dependent on Zookeeper being in a multi-node configuration, right? So it will be impossible to run OpenVDC on 1 machine?

File: ci/citest/acceptance-test/multibox/config.source [hunk 2]
```diff
@@ -16,6 +16,11 @@ PREFIX="24"
 BOXES=(
     "minimal-7.2.1511-x86_64.kvm.box"
 )
+ZK_NODES=(
+  "10.0.100.10"
+  "10.0.100.11"
+  "10.0.100.12"
```
Comment (Metallion) [inline ci/citest/acceptance-test/multibox/config.source:22]: It looks like all this array is used for is to output the list of zookeeper nodes when running the test. I think this is not a good way to do things. It needlessly clutters the configuration file.

The drawing in the illustrations folder should be enough to show where zookeeper is running.

**Review by Metallion — CHANGES_REQUESTED** [review]

File: ci/citest/acceptance-test/multibox/10.0.100.10-zookeeper/guestroot/var/lib/zookeeper/myid (hunk 1 above)
Comment (unakatsuo) [reply ci/citest/acceptance-test/multibox/10.0.100.10-zookeeper/guestroot/var/lib/zookeeper/myid:1]: https://zookeeper.apache.org/doc/r3.4.6/zookeeperAdmin.html#sc_zkMulitServerSetup

File: ci/citest/acceptance-test/multibox/config.source (hunk 2 above)
Comment (unakatsuo) [reply ci/citest/acceptance-test/multibox/config.source:22]: srvr Result from all the nodes is important for tracking the issue for cluster mode zookeeper. Especially ``Mode`` entry.

**Comment by unakatsuo** [comment]
It's still possible to run everything on one machine.

The reasons why I enabled the multi node configuration in this change are:

1. Reproduce the inconsistent read from zookeeper.
2. Make multibox CI equal to possible production configuration.

**Comment by unakatsuo** [comment]
Will update the diagram later.

File: ci/citest/acceptance-test/multibox/config.source (hunk 2 above)
Comment (Metallion) [reply ci/citest/acceptance-test/multibox/config.source:22]: Sorry, I misread the code. I saw the echo but missed the nc command. My bad.

~~Could I read up anywhere about what this sending "srvr" to tcp port 2181 does exactly?~~ Reading about it here atm https://zookeeper.apache.org/doc/r3.3.1/zookeeperAdmin.html

**Comment by Metallion** [comment]
Thanks for the clarification on the multi node configuration. Waiting for the diagram update. ^_^

File: ci/citest/acceptance-test/multibox/config.source (hunk 2 above)
Comment (Metallion) [reply ci/citest/acceptance-test/multibox/config.source:22]: Ok I see the output in the CI. This can stay. :)

**Review by Metallion — APPROVED** [review]

**Comment by Metallion** [comment]
Looks good to me. :+1: Shall I merge or do you still want that review from @b0r6 first?

**Pull request merged**